tensorflow/tools/pip_package/setup.py

```diff
@@ -77,8 +77,9 @@
 REQUIRED_PACKAGES = [
     # NOTE: As numpy has releases that break semver guarantees and several other
     # deps depend on numpy without an upper bound, we must install numpy before
-    # everything else.
-    'numpy ~= 1.19.2',
+    # everything else. So make sure that the correct version of NumPy
+    # is already installed before calling pip or setup.py
+    'numpy',
```

Comment:
I'd make this `>=1.19.2,<1.21.2` because it looks like the maintainers do not want to set the version constraint to include not-yet-released versions here.

Reply:
Or actually, `<1.22.0` would be better.

Reply:
Given the comment above, such a declaration would be only a check that the already installed version is correct. But if it is the only thing that blocks the fix, I am happy do include it.

Reply:
Not quite sure what you mean. This ends up in `install_requires`, which will make `pip` respect the constraint:
```python
setup(
   ...
    install_requires=REQUIRED_PACKAGES,
```
The "is already installed" in the comment is confusing/wrong, there is no already-installed version before invoking pip - where could it have come from?

Reply:
If I understand the comment and CI install scripts correctly, this alludes to the fact that you need to install numpy through pip or from other sources (e.g. apt-get) *before* you call  `pip install tensorflow`. Ugly, yes. Welcome to the fun of mixing different requirements resolvers. Nothing to fix within the scope of this PR.

Reply:
Oh, and look at https://github.com/tensorflow/tensorflow/commit/fe2fad5184f2ded836ccfeaa39b557472b69ce50#r52001636 to see where this came from.

Reply:
> If I understand the comment and CI install scripts correctly, this alludes to the fact that you need to install numpy through pip or from other sources (e.g. apt-get) before you call pip install tensorflow. Ugly, yes. Welcome to the fun of mixing different requirements resolvers

That's not the case:

```bash
$ conda create -n tf-test
$ conda activate tf-test
$ conda install python=3.8
$ pip install tensorflow
...
numpy                   1.19.5
...
tensorflow              2.5.0
```
Full output of `pip list`:

<details>

```
$ pip list
Package                 Version
----------------------- -------------------
absl-py                 0.13.0
astunparse              1.6.3
cachetools              4.2.2
certifi                 2021.5.30
charset-normalizer      2.0.4
flatbuffers             1.12
gast                    0.4.0
google-auth             1.34.0
google-auth-oauthlib    0.4.5
google-pasta            0.2.0
grpcio                  1.34.1
h5py                    3.1.0
idna                    3.2
keras-nightly           2.5.0.dev2021032900
Keras-Preprocessing     1.1.2
Markdown                3.3.4
numpy                   1.19.5
oauthlib                3.1.1
opt-einsum              3.3.0
pip                     21.2.2
protobuf                3.17.3
pyasn1                  0.4.8
pyasn1-modules          0.2.8
requests                2.26.0
requests-oauthlib       1.3.0
rsa                     4.7.2
setuptools              49.6.0.post20210108
six                     1.15.0
tensorboard             2.5.0
tensorboard-data-server 0.6.1
tensorboard-plugin-wit  1.8.0
tensorflow              2.5.0
tensorflow-estimator    2.5.0
termcolor               1.1.0
typing-extensions       3.7.4.3
urllib3                 1.26.6
Werkzeug                2.0.1
wheel                   0.36.2
wrapt                   1.12.1
```

</details>

There's only one resolver here, namely `pip`'s one.

Reply:
> Nothing to fix within the scope of this PR.

Agreed. Unless you wanted to delete the comment, it's just wrong. But I guess only do that if a maintainer asks for it.

Reply:
Ok then as noted in https://github.com/tensorflow/tensorflow/commit/fe2fad5184f2ded836ccfeaa39b557472b69ce50#r52001636 two months ago the comment is completely out of line, but no reaction to my remark, except for  https://github.com/tensorflow/tensorflow/pull/48935#issuecomment-858935936 which does not even address the comment. So I thought I clarify the comment in the setup.py and let the maintainers resolve their inconsitency on their own.

To be honest, I have little motivation left to contribute to TF any further. There is no support from Google or any (open source TF maintainer) to get any progress on this. They can either take the PR or leave it alone.

Reply:
Fair enough. Thanks for working on this, much appreciated. Let's see if we get a response here, I would like to see it merged - if only to help the NumPy/SciPy users who want an up-to-date version of both TF and NumPy.

Reply:
> no support form Google..
TF team has been through multiple reorgs over the past year and there has been strategy shifts regarding prioritization of OSS. Happy to say that support is coming soon and we should be able to handle these issues faster.

> up-to-date version of both TF and NumPY
We're working on this, but currently our CI strategy has been to use the exact same set of dependencies in all supported versions of Python and both in CI env and in `setup.py`. We have roadmap to decouple these but it would take a few quarters to fully land, unfortunately.

We can discuss more of this in SIG Build meetings, if needed.

Reply:
Thanks @mihaimaruseac, great to hear more priority will be given to this. I don't know what SIG Build meetings are, but please feel  free to ping me on NumPy versioning and wheels/conda dependency versioning topics.

Reply:
bit.ly/tf-sig-build-notes has all the details about SIG Build. Basically, it's a group of people interested in building TF on different environments